build/linter/lll/analyzer.go

```diff
@@ -0,0 +1,156 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lll
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"strings"
+	"unicode/utf8"
+
+	"golang.org/x/tools/go/analysis"
+)
+
+const lllName = "lll"
+
+const goCommentDirectivePrefix = "//go:"
+
+type settings struct {
+	// LineLength is the maximum line length.
+	LineLength int
+	// TabWidth is the width of a tab character.
+	TabWidth int
+}
+
+// Analyzer is the analyzer struct of lll.
+var Analyzer = &analysis.Analyzer{
+	Name: lllName,
+	Doc:  "Reports long lines",
+	Run: func(pass *analysis.Pass) (any, error) {
+		err := runLll(pass, &settings{
+			LineLength: 120,
+			TabWidth:   1,
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		return nil, nil
+	},
+}
+
+type result struct {
+	Filename string
+	Line     int
+	Text     string
+}
+
+func runLll(pass *analysis.Pass, settings *settings) error {
+	fileNames := make([]string, 0, len(pass.Files))
+	for _, f := range pass.Files {
+		pos := pass.Fset.PositionFor(f.Pos(), false)
+		if pos.Filename != "" && !strings.HasSuffix(pos.Filename, "failpoint_binding__.go") {
+			fileNames = append(fileNames, pos.Filename)
+		}
+	}
+
+	spaces := strings.Repeat(" ", settings.TabWidth)
+
+	for idx, f := range fileNames {
+		lintIssues, err := getLLLIssuesForFile(f, settings.LineLength, spaces)
+		if err != nil {
+			return err
+		}
+
+		for _, i := range lintIssues {
+			pass.Reportf(pass.Files[idx].Pos(), "%s:%d too long", i.Filename, i.Line)
+		}
+	}
+
+	return nil
+}
+
+func getLLLIssuesForFile(filename string, maxLineLen int, tabSpaces string) ([]result, error) {
+	var res []result
+
+	//nolint: gosec
+	f, err := os.Open(filename)
+	if err != nil {
+		return nil, fmt.Errorf("can't open file %s: %s", filename, err)
+	}
+	defer f.Close()
+
+	lineNumber := 0
+	multiImportEnabled := false
+
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		lineNumber++
+
+		line := scanner.Text()
+		line = strings.ReplaceAll(line, "\t", tabSpaces)
+
+		if strings.HasPrefix(line, goCommentDirectivePrefix) {
+			continue
+		}
+
+		if strings.HasPrefix(line, "import") {
```

Comment:
can use related position of ast.File.Imports to skip import part

Reply:
Done